Source/WebDriver/soup/WebSocketServerSoup.cpp

```diff
@@ -0,0 +1,208 @@
+        soup_websocket_connection_send_message(message.connection.get(), SOUP_WEBSOCKET_DATA_TEXT, rawMessage.get());
+    });
+}
+
+static void handleWebSocketConnection(SoupServer*, SoupServerMessage*, const char* path, SoupWebsocketConnection* connection, gpointer userData)
+{
+    // Delayed steps from the end of the handshake, as now we have a connection object
+    auto webSocketServer = static_cast<WebSocketServer*>(userData);
+    if ("/session"_s == path)
+        webSocketServer->addStaticConnection({ connection });
+    else {
+        auto resourceName = String::fromLatin1(path);
+        auto sessionId = webSocketServer->getSessionID(resourceName);
+        auto connectionForKey = WebSocketMessageHandler::Connection(connection);
+        webSocketServer->addConnection(connectionForKey, sessionId);
+    }
+
+    g_signal_connect(connection, "closed", G_CALLBACK(+[](SoupWebsocketConnection* connection, gpointer userData) {
+        auto webSocketServer = static_cast<WebSocketServer*>(userData);
+        webSocketServer->messageHandler().clientDisconnected(connection);
+    }), webSocketServer);
+
+    g_signal_connect(connection, "message", G_CALLBACK(handleWebSocketMessage), webSocketServer);
+}
+
+
+std::optional<String> WebSocketServer::listen(const String& host, unsigned port)
+{
+#if USE(SOUP2)
+    WTFLogAlways("WebSockets support not implemented yet with libsoup2");
+    return stf::nullopt;
+#endif
+
+    m_soupServer = adoptGRef(soup_server_new("server-header", "WebKitWebDriver-WSS", nullptr));
+    GUniqueOutPtr<GError> error;
+    if (!soupServerListen(m_soupServer.get(), host, port, &error.outPtr())) {
+        WTFLogAlways("Failed to start WebSocket server at port %u: %s", port, error->message);
+        return std::nullopt;
+    }
+
+    // Callback for handling incoming WebSocket handshake requests
+    soup_server_add_handler(m_soupServer.get(), nullptr, handleIncomingHandshake, this, nullptr);
+
+    // Callback called when the WebSocket handshake has finished and we have a valid WebSocket Connection object
+    soup_server_add_websocket_handler(m_soupServer.get(), nullptr, nullptr, nullptr, handleWebSocketConnection, this, nullptr);
+
+    // "/session" is the default resource to start bidi-only sessions
+    m_listener = WebSocketListener::create(
+        host.isNull() ? "localhost"_s : host,
+        port,
+        false,
+        { "/session"_s }
+    );
+    return getWebSocketURL(m_listener, nullString());
+}
+
+void WebSocketServer::sendMessage(const String& session, const String& message)
+{
+    for (const auto& pair : m_connectionToSession) {
+        if (pair.value == session) {
+            GRefPtr<GBytes> rawMessage = adoptGRef(g_bytes_new(message.utf8().data(), message.length()));
+            soup_websocket_connection_send_message(pair.key.get(), SOUP_WEBSOCKET_DATA_TEXT, rawMessage.get());
+            return;
+        }
+    }
+}
```

Comment:
Where is this used?

Reply:
This was a leftover from splitting the command dispatching parts into a follow-up patch. I've moved it there.